src/jit/emitxarch.cpp

```diff
@@ -2847,15 +2847,34 @@ regNumber emitter::emitInsBinary(instruction ins, emitAttr attr, GenTree* dst, G
     assert(!dst->isContained() ||
            dst->isContainedIndir() ||
            dst->isContainedLclField() ||
+           dst->isContainedLclVar() ||  // this is essentially a contained LclField with zero offset
            instrIs3opImul(ins)); // dst on these isn't really the dst
 
+#ifdef DEBUG
     // src can be anything but both src and dst cannot be addr modes
     // or at least cannot be contained addr modes
     if (dst->isContainedIndir())
+    {
         assert(!src->isContainedIndir());
+    }
+
+    if (dst->isContainedLclVar())
+    {
+        assert(!src->isContainedIndir());
+        assert(!src->isContainedLclField());
+        assert(!src->isContainedLclVar());
+    }
 
     if (src->isContainedLclField())
+    {
         assert(!dst->isContained());
+    }
+
+    if (src->isContainedLclVar())
+    {
+        assert(!dst->isContained());
+    }
+#endif
 
```

Comment:
We already have GenTree::isContainedMemoryOp() defined as

`return isContained() && isMemoryOp();`

I have modified it as follows

`return (isContained() && isMemoryOp()) || isContainedLclVar();`

Further, all of the above asserts in emitxarch.cpp reduce to

```
    if (dst->isContainedMemoryOp())
    {
        assert(!src->isContainedMemoryOp());
    }

    if (src->isContainedMemoryOp())
    {
        assert(!dst->isContainedMemoryOp());
    }

```